Solana CCIP integration tests

archseer opened the pull request:
<!--- Does this work have a corresponding ticket?  Please link it here as well as one of:
    - the PR title
    - branch name
    - commit message
[LINK-777](https://smartcontract-it.atlassian.net/browse/LINK-777)
--> 

### Requires
<!--- Does this work depend on other open PRs? Please list them.
- https://github.com/smartcontractkit/chainlink-common/pull/7777777
-->

### Supports
<!--- Does this work support other open PRs?  Please list them.
- https://github.com/smartcontractkit/ccip/pull/7777777
-->

archseer committed:
ccip_messaging_test: Make messagingTestCase receiver chain agnostic

archseer committed:
Test_CCIPMessaging_Solana skeleton

archseer committed:
Draft Solana versions of commit and exec confirmers

archseer committed:
extract sourceDest

archseer committed:
Implement getLatestNonce for Solana

archseer committed:
Implement WaitForTheToken balance for Solana

archseer committed:
Update ConfirmMultipleCommits with Solana support

archseer committed:
wip

archseer committed:
Include SVM extra args

archseer committed:
wip

archseer committed:
fix for startBlock(s) potentially being nil

archseer committed:
fix SVM extra args serialization

archseer committed:
only replay blocks on EVM chains, this is EVM specific

archseer committed:
Add a LatestBlock helper

archseer committed:
Fix post-rebase

archseer committed:
Refactor: use []byte for addresses when asserting balances

archseer committed:
Add TestTokenTransfer_Solana

archseer committed:
nonce account may not exist yet

archseer committed:
TEMP: disable Solana token transfer test until code is merged

archseer committed:
fix: Use correct chain family when fetching CR config

archseer committed:
oraclecreator: Only initialize CR/CW instances for relayers with CCIP support

archseer committed:
Rename DoSendRequest to SendRequest

archseer committed:
Avoid panic on solana getting passed EVm offramp address

archseer committed:
Don't fail if the block doesn't contain that event type

archseer committed:
solana: Handle OCR3 producing empty reports

prashantkumar1982 commented on core/capabilities/ccip/ccipsolana/executecodec.go:
```diff
@@ -26,6 +26,13 @@ func NewExecutePluginCodecV1() *ExecutePluginCodecV1 {
 }
 
 func (e *ExecutePluginCodecV1) Encode(ctx context.Context, report cciptypes.ExecutePluginReport) ([]byte, error) {
+	if len(report.ChainReports) == 0 {
```
Same kind of change is needed on the ccipsolana/commitcodec.go file too right?
cc @huangzhen1997

tt-cll committed:
image hotfix

ilija42 committed:
Fix codec modifiers for OffRamp Source Chain Config

archseer committed:
Allow commit with price updates only, no merkle roots

ilija42 committed:
Fix MethodNameGetLatestPriceSequenceNumber cfg

archseer committed:
Disable RMN verification on Solana destination

archseer committed:
CR needs to read both Config and State properties of the config

archseer committed:
Configure TokenInfo for Solana

amit-momin committed:
Updated ChainConfigGasPrice account config

archseer committed:
Fix commit price only report encoding

archseer committed:
solana: Fix OCR3 report hashing

archseer committed:
environment/memory: Fund the solana transmitter keys

archseer committed:
fix ChainConfigGasPriceConfig lookup

archseer committed:
correct IndexedField0

archseer committed:
Read events from offramp, not router

archseer committed:
allow v0 transactions

archseer committed:
skip price-only events

archseer committed:
Fix config for GetTokenPrice

archseer committed:
Extract the token field

archseer committed:
solana: Set gas helpers to non-zero to appease the CCIP plugin

archseer committed:
Fix the LinkTokenMint property

archseer committed:
Return router as the nonce manager

archseer committed:
Use a custom workaround in chainlink-solana to return the right config based on pluginType

archseer committed:
Set a non-zero execution fee estimate

archseer committed:
ccipsolana: Add OnRampAddress and MerkleRoot to report

Requires changes in chainlink-ccip

archseer committed:
solana: fix GetTokenPrice for source chain reader too

archseer committed:
Update OnRampAddress and Root handling to match upstream contract changes

archseer committed:
fix compatibility with latest chainlink-ccip again

https://github.com/smartcontractkit/chainlink-ccip/commit/d3fde0a866bd03b451b2e0f873aa80edffa84abb

archseer committed:
fix nonce asserts

archseer committed:
fix compat with latest upstream again

archseer committed:
TokenTransfer test

archseer committed:
Adjust assertions in messaging test

archseer committed:
Add a blank RMNRemote definition to work with latest contracts

archseer committed:
solana: Add DeployerKey to allowed senders

archseer committed:
Support solana in AddLanesForAll

ilija42 committed:
Add rmn contract reader cfgs

archseer committed:
Refactor SendRequest to support Solana

archseer committed:
Add a Solana -> EVM test case

archseer committed:
Defince CrossChainAmount type, CCIPMessageSent event

archseer committed:
Adjust GetWrappedNative definition

archseer committed:
GetDynamicConfig needs to be nested under DynamicConfig

archseer committed:
Properly encode Solana addresses in config tracker

archseer committed:
Since the home config is using EIP-55 now, address codec can do the same

archseer committed:
ccipevm: Pad DestTokenAddress to make it ABI decodable

archseer committed:
ccipsolana: Avoid modifying raw data when converting endianness

archseer committed:
TEMP: work around 32 byte values returned rather than u32

archseer committed:
revert-reason: Handle TokenHandlingError having two values

archseer committed:
Add SVM->EVM token transfer test, fix transfer helpers to support it

archseer committed:
small tweak

archseer committed:
tests: Change the decimals value for the remote token

archseer committed:
Cleanup tests and fix assertions on decimals

archseer committed:
Upstream contract was updated to use a u32_be for this value

archseer committed:
match value to upstream tests now that the juels amounts are correct

archseer committed:
Remove CW config that's now part of transform in chainlink-solana

archseer committed:
remove stray printf

archseer committed:
solana: Only one OnRamp address is now stored in config

archseer committed:
clean up comments on OffRampLatestConfigDetails

archseer committed:
Fix test compilation

archseer committed:
Address lints

archseer committed:
Don't require.NoError in goroutine

archseer committed:
Test execution with extra accounts

archseer committed:
go mod tidy

archseer committed:
Use a fork of the solana binary library, gomods tidy

archseer committed:
Update revert-reason

archseer committed:
fix more lints

archseer committed:
avoid assert in goroutine

amit-momin commented on core/capabilities/ccip/configs/solana/chain_writer.go:
```diff
@@ -244,70 +244,6 @@ func getExecuteMethodConfig(fromAddress string, offrampProgramAddress string) ch
-					IncludeIndexes:  []int{},
-				},
-				Optional: true,
-			},
```
Could we add a comment here mentioning that these accounts for token transfers are added during the CCIPExecute args transform? Something like:
```
// user token account, token billing config, pool chain config, and pool lookup table accounts are added to the accounts list in the CCIPExecute args transform for each token transfer
```

archseer committed:
fix more lints and tests

archseer commented on core/capabilities/ccip/ccipevm/addresscodec.go:
```diff
@@ -4,13 +4,14 @@ import (
 func (a AddressCodec) AddressBytesToString(addr []byte) (string, error) {
-	// TODO support EIP-55 checksum, https://smartcontract-it.atlassian.net/browse/CCIP-5340
-	return "0x" + hex.EncodeToString(addr), nil
+	return gethcommon.BytesToAddress(addr).Hex(), nil
```
Note: this will require an update on the EVM staging cluster

huangzhen1997 replied:
If we are going to support EIP-55, maybe we should update the `AddressStringToBytes()` to use geth common library as well.

huangzhen1997 replied:
Also adding a length check for the address, since `BytesToAddress` doesn't check it. Check [this](https://github.com/smartcontractkit/chainlink/commit/82fade97ba57a8242abcf0c7d828f6137b17dd76#diff-62c6e868fa1f049d352664bd255091657013ada17214f1bd807f57f848547b15) as an example

huangzhen1997 replied:
check [this](https://github.com/smartcontractkit/chainlink/commit/82fade97ba57a8242abcf0c7d828f6137b17dd76#diff-62c6e868fa1f049d352664bd255091657013ada17214f1bd807f57f848547b15) as an example

archseer commented on core/capabilities/ccip/ccipsolana/gas_helpers.go:
```diff
@@ -13,10 +13,10 @@ type EstimateProvider struct {
 // CalculateMerkleTreeGas is not implemented
 func (gp EstimateProvider) CalculateMerkleTreeGas(numRequests int) uint64 {
-	return 0
+	return 1
```
I can't remember why this was necessary but the plugin didn't like it when the cost estimate ended up being zero

huangzhen1997 replied:
It's used when verify report, making sure the estimated total gas below max Gas. @winder I remembered we discussed this before, setting to 0 to indicate noop for Solana gas estimate provider. Is it okay to set it to 1 ?

makramkd replied:
Do we have the logs / errors for when this was set to 0?

It doesn't matter too much for Solana since we'll always execute 1msg/report; ideally that'd be clear when looking through this gashelpers code.

winder replied:
I agree with @makramkd here. 1 is fine. Perhaps things could be improved in the future with a `NoOp` provider that can be shared with most/all of the non-EVMs.

archseer committed:
Fix some of the Messaging_EVM2Solana test failure

huangzhen1997 reviewed (commented):
If we are going to support EIP-55, maybe we should update the `AddressStringToBytes()` to use geth common library as well. Check [this](https://github.com/smartcontractkit/chainlink/commit/82fade97ba57a8242abcf0c7d828f6137b17dd76#diff-62c6e868fa1f049d352664bd255091657013ada17214f1bd807f57f848547b15) as an example

amit-momin committed:
Merge branch 'develop' into solana-messaging-test

amit-momin committed:
Tidied go mod

amit-momin committed:
Fixed lint error

huangzhen1997 commented:
@amit-momin Do you want to update the `StartIndex` in the chain_writer.go as well ? 
`IsWritable: chainwriter.MetaBool{BitmapLocation: "ExtraData.ExtraArgsDecoded.accountIsWritableBitmap", StartIndex: 1},`

amit-momin commented:
> @amit-momin Do you want to update the `StartIndex` in the chain_writer.go as well ? `IsWritable: chainwriter.MetaBool{BitmapLocation: "ExtraData.ExtraArgsDecoded.accountIsWritableBitmap", StartIndex: 1},`

Yup I'll get that added too but waiting to pin the latest `chainlink-solana` dependency first

amit-momin committed:
Merge branch 'develop' into solana-messaging-test

amit-momin committed:
Updated Solana ChainWriter execute user accounts config

amit-momin committed:
Addressed feedback

amit-momin committed:
Fixed test error

amit-momin committed:
Merge branch 'develop' into solana-messaging-test

amit-momin committed:
Updated SVM transmitter commit price-only condition

amit-momin committed:
Merge branch 'develop' into solana-messaging-test

amit-momin committed:
Fixed lint error

archseer commented:
https://github.com/smartcontractkit/chainlink/pull/15965/commits/9f63e286e59e0de9fb5ffca561d8c002a30c93b3 seems to have broken tests across the board, likely because all these addresses are padded to 32 bytes, not 20 bytes (AddressLength).

archseer committed:
Revert 9f63e28 partially, addresses are sometimes padded to 32 bytes

archseer committed:
Move helpers to test_helpers to fix running the token transfer test

SerializeSVMExtraArgs was defined in the messaging test, so it wasn't
available in the token transfer test for some reason

archseer committed:
Merge branch 'develop' into solana-messaging-test

archseer committed:
gomods tidy

archseer committed:
fix compilation

archseer committed:
Fix TokenAdminRegistry definition (it's now under ccipCommon)

archseer committed:
Fix assert on counter state

archseer committed:
Disable addresscodec test that's currently not applicable

archseer committed:
Merge branch 'develop' into solana-messaging-test

archseer committed:
goimports

Galaxysurfer2034 commented:
28 issues?? Remember that one time when Solana was working perfectly and at
optimal quality?.. yeah me neither. Lmao

On Tue, Mar 25, 2025, 12:33 AM cl-sonarqube-production[bot] <
***@***.***> wrote:

> [image: Quality Gate passed]
> <https://sonarqube.main.prod.cldev.sh/dashboard?id=smartcontractkit_chainlink&pullRequest=15965> *Quality
> Gate passed*
>
> Issues
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&resolved=false> 46
> New issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&resolved=false>
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&fixedInPullRequest=15965> 28
> Fixed issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&fixedInPullRequest=15965>
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&issueStatuses=ACCEPTED> 0
> Accepted issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&issueStatuses=ACCEPTED>
>
> Measures
>
> <https://sonarqube.main.prod.cldev.sh/security_hotspots?id=smartcontractkit_chainlink&pullRequest=15965> 0
> Security Hotspots
> <https://sonarqube.main.prod.cldev.sh/security_hotspots?id=smartcontractkit_chainlink&pullRequest=15965>
>
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_coverage&view=list> 0.0%
> Coverage on New Code
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_coverage&view=list>
>
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_duplicated_lines_density&view=list> 1.4%
> Duplication on New Code
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_duplicated_lines_density&view=list>
>
> See analysis details on SonarQube
> <https://sonarqube.main.prod.cldev.sh/dashboard?id=smartcontractkit_chainlink&pullRequest=15965>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/smartcontractkit/chainlink/pull/15965#issuecomment-2750136402>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BJR4SADGLIKTCVFDFB3UJO32WDTA7AVCNFSM6AAAAABVLLUBA6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJQGEZTMNBQGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
> [image: cl-sonarqube-production[bot]]*cl-sonarqube-production[bot]* left
> a comment (smartcontractkit/chainlink#15965)
> <https://github.com/smartcontractkit/chainlink/pull/15965#issuecomment-2750136402>
> [image: Quality Gate passed]
> <https://sonarqube.main.prod.cldev.sh/dashboard?id=smartcontractkit_chainlink&pullRequest=15965> *Quality
> Gate passed*
>
> Issues
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&resolved=false> 46
> New issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&resolved=false>
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&fixedInPullRequest=15965> 28
> Fixed issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&fixedInPullRequest=15965>
>
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&issueStatuses=ACCEPTED> 0
> Accepted issues
> <https://sonarqube.main.prod.cldev.sh/project/issues?id=smartcontractkit_chainlink&pullRequest=15965&issueStatuses=ACCEPTED>
>
> Measures
>
> <https://sonarqube.main.prod.cldev.sh/security_hotspots?id=smartcontractkit_chainlink&pullRequest=15965> 0
> Security Hotspots
> <https://sonarqube.main.prod.cldev.sh/security_hotspots?id=smartcontractkit_chainlink&pullRequest=15965>
>
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_coverage&view=list> 0.0%
> Coverage on New Code
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_coverage&view=list>
>
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_duplicated_lines_density&view=list> 1.4%
> Duplication on New Code
> <https://sonarqube.main.prod.cldev.sh/component_measures?id=smartcontractkit_chainlink&pullRequest=15965&metric=new_duplicated_lines_density&view=list>
>
> See analysis details on SonarQube
> <https://sonarqube.main.prod.cldev.sh/dashboard?id=smartcontractkit_chainlink&pullRequest=15965>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/smartcontractkit/chainlink/pull/15965#issuecomment-2750136402>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BJR4SADGLIKTCVFDFB3UJO32WDTA7AVCNFSM6AAAAABVLLUBA6VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONJQGEZTMNBQGI>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

makramkd commented on core/capabilities/ccip/ccipevm/addresscodec.go:
```diff
@@ -1,23 +1,20 @@
 func (a AddressCodec) AddressBytesToString(addr []byte) (string, error) {
-	// TODO support EIP-55 checksum, https://smartcontract-it.atlassian.net/browse/CCIP-5340
-	return "0x" + hex.EncodeToString(addr), nil
+	return common.BytesToAddress(addr).Hex(), nil
```
So what was the root issue here?

winder replied:
Pretty sure this was just because we can. We would have implemented it like this in chainlink-ccip if we had access to go-ethereum.

huangzhen1997 replied:
And to support the checksum hex align with EIP-55

makramkd replied:
Didn't this break things before? That's why we had the ticket right?

archseer replied:
Yeah it required the change in deployments tooling so that the onchain state is configured correctly. Since I fixed that at the same time it's no longer broken, but it does mean that you'll have to update the onchain DON config on your staging deployment

makramkd commented on core/capabilities/ccip/ccipevm/msghasher.go:
```diff
@@ -113,7 +114,7 @@ func (h *MessageHasherV1) Hash(ctx context.Context, msg cciptypes.Message) (ccip
 		lggr.Debugw("decoded dest gas amount",
 			"destGasAmount", destGasAmount)
 
-		destTokenAddress, err := abiDecodeAddress(rta.DestTokenAddress)
+		destTokenAddress, err := abiDecodeAddress(common.LeftPadBytes(rta.DestTokenAddress, 32))
```
I feel like this may end up masking some issues - the contracts have to emit something that must be abi-encoded if the destination is EVM.

prashantkumar1982 replied:
I too have a question about this.
In SVM -> EVM flow, the incoming rta.DestTokenAddress won't be abi encoded right?
Why are we calling abiDecodeAddress() here irrespective of source family?

Or do we expect the onchain user on source chain to provide an ABI encoded []byte for destination addresses?

huangzhen1997 replied:
how about use `common.BytesToAddress(tokenAmount.SourcePoolAddress)`, same with the [executecodec](https://github.com/smartcontractkit/chainlink/blob/a9ccdda94a09549c8d720fa07828603bab3855b7/core/capabilities/ccip/ccipevm/executecodec.go#L99)

makramkd replied:
> Or do we expect the onchain user on source chain to provide an ABI encoded []byte for destination addresses?

The onchain user _must_ specify a valid receiver address for the dest chain family. This is enforced on EVM -> EVM and EVM -> SOL, should be enforced the other way around as well.

makramkd replied:
If we don't have this kind of behavior & validation in solana then I believe this is a bug and should be fixed?

prashantkumar1982 replied:
@makramkd : How could we validate on source chain, whether destination address is valid?

archseer replied:
We specify a valid address, but we specify it as 20 bytes, not the fully padded 32 bytes

winder replied:
I have the same question as @prashantkumar1982 and @huangzhen1997

It seems like this needs some followup to make sure this handles all families with the address codec.

Could we add a TODO or create a followup ticket?

makramkd commented on core/capabilities/ccip/ccipsolana/commitcodec.go:
```diff
@@ -186,13 +192,17 @@ func encodeBigIntToFixedLengthLE(bi *big.Int, length int) []byte {
+
 	// Reverse the byte array to convert it from little-endian to big-endian
-	for i, j := 0, len(data)-1; i < j; i, j = i+1, j-1 {
-		data[i], data[j] = data[j], data[i]
+	for i, j := 0, len(buf)-1; i < j; i, j = i+1, j-1 {
+		buf[i], buf[j] = buf[j], buf[i]
 	}
```
Can use [slices.Reverse](https://pkg.go.dev/slices#Reverse).

makramkd commented on core/capabilities/ccip/ccipsolana/executecodec.go:
```diff
@@ -31,6 +31,13 @@ func NewExecutePluginCodecV1(extraDataCodec common.ExtraDataCodec) *ExecutePlugi
 
 func (e *ExecutePluginCodecV1) Encode(ctx context.Context, report cciptypes.ExecutePluginReport) ([]byte, error) {
+	if len(report.ChainReports) == 0 {
+		// OCR3 runs in a constant loop and will produce empty reports, so we need to handle this case
+		// return an empty report, CCIP will discard it on ShouldAcceptAttestedReport/ShouldTransmitAcceptedReport
+		// via validateReport before attempting to decode
```
While this is true and this is fine as a sanity check, if this gets called with `ChainReports` being empty there is a bug in cl-ccip somewhere. cc @winder

prashantkumar1982 replied:
I remember @archseer said previously that he sees ChainReports being empty in e2e testing, and thus had to fix this code.

winder replied:
AFAIK there was a bug, and it has been fixed.

makramkd commented on core/capabilities/ccip/ccipsolana/extradatadecoder.go:
```diff
@@ -74,17 +74,7 @@ func (d ExtraDataDecoder) DecodeExtraArgsToMap(extraArgs cciptypes.Bytes) (map[s
-	return map[string]interface{}{
-		svmDestExecDataKey: bytesToUint32LE(destExecData),
+	return map[string]any{
+		svmDestExecDataKey: binary.BigEndian.Uint32(destExecData),
```
This changes semantics, previous function was using LittleEndian, intended?

huangzhen1997 replied:
Have the same question. The extraDataCodec should be source chain specific, for Solana on-chain encoding use little-endian

prashantkumar1982 replied:
Yes, this seems wrong. For Solana, data would be encoded as LE.

archseer replied:
Contracts are using big endian: https://github.com/smartcontractkit/chainlink-ccip/commit/7dc14d5baa62569b687b829171463aeeda91f1a6

Maybe we should change it to little endian, \cc @PabloMansanet

winder replied:
Looks like this detail may need some followup?

archseer replied:
We'll keep the same endianness here

makramkd commented on core/capabilities/ccip/oraclecreator/plugin.go:
```diff
@@ -608,11 +613,13 @@ func createChainWriter(
-		if solana.PublicKeyLength != len(offrampProgramAddress) {
-			return nil, fmt.Errorf("invalid offrampProgramAddress length: %d", len(offrampProgramAddress))
+		var offrampProgramAddress solana.PublicKey
+		// NOTE: this function can still be called with EVM inputs, and PublicKeyFromBytes will panic on addresses with len=20
+		// technically we only need the writer to do fee estimation so this doesn't matter and we can use a zero address
```
I guess this is a bit counter-intuitive, worth a potential re-organization of this code in future PRs.

archseer replied:
Yeah I agree, we were talking about extracting EstimateFee from chain writer. We could just always bind the zero address here since it doesn't really matter what the value is. On EVM we end up binding it to the address from the destination chain which doesn't exist on the source chain at all

makramkd commented on core/services/keystore/keys/ocr2key/solana_keyring.go:
```diff
@@ -48,15 +50,22 @@ func (skr *solanaKeyring) Sign(reportCtx ocrtypes.ReportContext, report ocrtypes
 }
 
 func (skr *solanaKeyring) Sign3(digest types.ConfigDigest, seqNr uint64, r ocrtypes.Report) (signature []byte, err error) {
```
What would be nice as a follow up is to have tests that sign reports w/ these funcs and verify the sigs using the contract code (similar to the current ContractTransmitter test for EVM)

makramkd commented on deployment/ccip/changeset/internal/deploy_home_chain.go:
```diff
@@ -409,9 +406,24 @@ func BuildOCR3ConfigForCCIPHome(
-			parsed, err2 := common.ParseHexOrString(string(transmitter))
-			if err2 != nil {
-				return nil, err2
+			// TODO: this should just use the addresscodec
```
Can we not create the address codec here? `deployment` is dependent on `chainlink` AFAIK

makramkd commented on deployment/ccip/changeset/internal/deploy_home_chain.go:
```diff
@@ -409,9 +406,24 @@ func BuildOCR3ConfigForCCIPHome(
+				if err != nil {
+					return nil, fmt.Errorf("failed to decode SVM address '%s': %w", transmitter, err)
+				}
+				parsed = pk.Bytes()
```
Just confirming, but this is the actual 32 byte public key right, not the string representation?

archseer replied:
Yep exactly 👍🏻

makramkd commented on deployment/ccip/changeset/testhelpers/messagingtest/helpers.go:
```diff
@@ -116,10 +127,16 @@ func getLatestNonce(tc TestCase) uint64 {
-		// err = common.GetAccountDataBorshInto(ctx, solanaGoClient, nonceEvmPDA, config.DefaultCommitment, &nonceCounterAccount)
-		// require.NoError(t, err, "failed to get account info")
-		// require.Equal(t, uint64(1), nonceCounterAccount.Counter)
+		ctx := context.Background()
```
nit: can use `tc.T.Context()` I think? If that `*testing.T` is available.

makramkd commented on deployment/ccip/changeset/testhelpers/messagingtest/helpers.go:
```diff
@@ -116,10 +127,16 @@ func getLatestNonce(tc TestCase) uint64 {
+		chainSelectorLE := solcommon.Uint64ToLE(tc.DestChain)
+		noncePDA, _, err := solana.FindProgramAddress([][]byte{[]byte("nonce"), chainSelectorLE, tc.Sender}, tc.OnchainState.SolChains[tc.DestChain].Router)
+		require.NoError(tc.T, err)
+		var nonceCounterAccount ccip_router.Nonce
+		// we ignore the error because the account might not exist yet
+		_ = solcommon.GetAccountDataBorshInto(ctx, client, noncePDA, solconfig.DefaultCommitment, &nonceCounterAccount)
+		latestNonce = nonceCounterAccount.Counter
```
Wait are nonces a thing on Solana? I thought everything was out of order execution? Would this always be zero?

archseer replied:
Yeah I added support to this function but then realized it doesn't make sense. I forget how this fits together but there are still nonce accounts for the other direction (`svm->` rather than `->svm`)

archseer replied:
I'll probably just replace that section with a no-op/fail

makramkd commented on integration-tests/smoke/ccip/ccip_messaging_test.go:
```diff
@@ -185,6 +195,204 @@ func Test_CCIPMessaging(t *testing.T) {
+		)
+	})
+
+	fmt.Printf("out: %v\n", out)
```
nit/ `t.Logf`

makramkd commented on integration-tests/smoke/ccip/ccip_token_transfer_test.go:
```diff
@@ -217,5 +360,182 @@ func TestTokenTransfer(t *testing.T) {
+		testhelpers.SeqNumberRangeToSlice(expectedSeqNums),
+		startBlocks,
+	)
+	// NOTE: can't validate this since there will be more execution states due to price updates being separate
```
What does ExecutionStateChanged have to do with price updates? Aren't those only emitted on msg executions?

archseer replied:
Hmm I think my comment is wrong here but for some reason the sequence was higher than `1` here for me (the tests for EVM assert seqNum = old num + 1). I think commit reports with price only updates still end up raising the seq number, so the total number was higher for me

archseer replied:
Hmm it seems to pass for me again so I'll re-enable

AnieeG commented on deployment/ccip/changeset/internal/deploy_home_chain.go:
```diff
@@ -409,9 +406,24 @@ func BuildOCR3ConfigForCCIPHome(
-			if err2 != nil {
-				return nil, err2
+			// TODO: this should just use the addresscodec
+			family, err := chain_selectors.GetSelectorFamily(destSelector)
```
Please create a helper

AnieeG commented on deployment/ccip/changeset/testhelpers/messagingtest/helpers.go:
```diff
@@ -95,12 +101,17 @@ type TestCaseOutput struct {
 }
 
-func sleepAndReplay(t *testing.T, e testhelpers.DeployedEnv, sourceChain, destChain uint64) {
+func sleepAndReplay(t *testing.T, e testhelpers.DeployedEnv, chainSelectors ...uint64) {
```
```suggestion
func sleepAndReplayOnEVM(t *testing.T, e testhelpers.DeployedEnv, chainSelectors ...uint64) {
```

archseer replied:
I'll keep the name and add a TODO, looks like this is getting implemented for solana: https://github.com/smartcontractkit/chainlink-solana/pull/1135

winder replied:
Looks like the linked PR has been merged. @AnieeG, does this address your concern?

archseer replied:
I've updated the code to use replay on Solana

AnieeG commented on deployment/ccip/changeset/testhelpers/solana/counter.go:
```diff
@@ -0,0 +1,54 @@
+package solana
```
should this be a common helper instead of testhelper

archseer replied:
I don't have a preference on the structure but this is parsing a receiver contract that shouldn't be used outside of tests

AnieeG commented on deployment/ccip/changeset/testhelpers/test_helpers.go:
```diff
@@ -154,7 +163,29 @@ func DeployTestContracts(t *testing.T,
 	}
 }
 
+func LatestBlock(ctx context.Context, env deployment.Environment, chainSelector uint64) (uint64, error) {
```
How about making it a method on deployment.Environment ?

AnieeG reviewed (commented):
I have done a partial review. As you mentioned, it will help with separate testhelpers of solana instead of having it within the same ones for EVM.

huangzhen1997 commented on core/capabilities/ccip/ccipevm/addresscodec_test.go:
```diff
@@ -24,34 +24,33 @@ func TestAddressStringToBytes(t *testing.T) {
 	require.Equal(t, want, got)
 }
 
+// we allow various sizes since some contracts store the 20-byte address as 32-byte
```
Maybe also add a comment in the AddressBytesToString()
 from `core/capabilities/ccip/ccipevm/addresscodec.go`, as the reason we skip the length check

prashantkumar1982 commented on core/capabilities/ccip/ccipevm/addresscodec_test.go:
```diff
@@ -24,34 +24,33 @@ func TestAddressStringToBytes(t *testing.T) {
 
+// we allow various sizes since some contracts store the 20-byte address as 32-byte
+// func TestInvalidAddressBytesToString(t *testing.T) {
+// 	addressCodec := AddressCodec{}
```
Can these be now removed?

archseer committed:
Only fetch the chainID in the EVM branch

Solana can use a string based chain ID

ilija42 committed:
Fix Solana CCIP ContractNameRMNProxy cfg

archseer committed:
fix some minor review comments

archseer committed:
Re-enable assert since code is passing again

archseer committed:
fix: Token and GasPrice data is stored big-endian encoded in FeeQuoter

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
Revert "Fix Solana CCIP ContractNameRMNProxy cfg"

This reverts commit 8a0652e927cf26cbe1c15a34e8c576f9e3599a36.

archseer committed:
gomods tidy

archseer committed:
Fix asserts with the new InProgress event present

archseer committed:
fix lint

archseer committed:
Correctly parse multiple emitted events in one tx

archseer committed:
Bump chainlink-ccip/chains/solana with the bug fix

archseer committed:
Bump again

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
Fix decoding CommitReportAccepted

ilija42 committed:
Solana Contract reader RMN MethodNameGetARM cfg changes

amit-momin commented on deployment/ccip/changeset/testhelpers/test_assertions.go:
```diff
@@ -389,6 +442,147 @@ func ConfirmCommitWithExpectedSeqNumRange(
+				return
+			case <-ticker.C:
+				// Scan for transactions referencing the address
+				ctx := context.Background()
```
nit: shouldn't we use `t.Context()` here instead?

amit-momin reviewed (dismissed):
Approving since ChainWriter relevant changes look good. Also the open threads seem like they can be addressed as follow up work/tickets.

archseer committed:
bump chainlink-solana

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
gomods tidy

archseer committed:
nix: Update flake

archseer committed:
Revert "Solana Contract reader RMN MethodNameGetARM cfg changes"

This reverts commit 4e828fecab9ffc27b9c6482965a52ba007a17c90.

archseer committed:
Replay on all chains, the new API supports non-EVM

archseer committed:
Fix issues with replay, make other methods work correctly

reductionista committed:
Use query_key_debugging branch of chainlink-solana

ilija42 committed:
Bump common and update ccip solana contract reader config

archseer committed:
Revert "Fix decoding CommitReportAccepted"

This reverts commit baf4628663e1f8f6f3a829e1a9c2d4b681634ed5.

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
fix build

archseer committed:
Disable chainID verification in e2e tests since we use a custom chainID

archseer committed:
Bump chainlink-solana to pull in fixes

archseer committed:
Override RMNRemote version so it's considered disabled

archseer committed:
Use a workaround for chainlink-ccip to avoid RMN issues

amit-momin committed:
Upgraded chainlink-solana dependency

archseer committed:
Disable standard capabilities build, repo is broken

archseer committed:
Only check root validation if there actually is a root

archseer committed:
Check for either nil or empty array

archseer committed:
Revert "Upgraded chainlink-solana dependency"

This reverts commit 8b4916bd0cdf88bb6ffb24db2db9965dc000441c.

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
Bump chainlink-ccip workaround branch again

archseer committed:
Fix the token overhead config in tests

archseer committed:
Merge remote-tracking branch 'origin/develop' into solana-messaging-test

archseer committed:
Revert "Disable standard capabilities build, repo is broken"

This reverts commit 7f96c92fccd2db9d05177398e135e559eb76d0da.

archseer committed:
Bump chainlink-solana, chainlink-ccip
